core/variant/variant_op.cpp

```diff
@@ -257,8 +257,13 @@ class OperatorEvaluatorShiftLeft {
 	static void evaluate(const Variant &p_left, const Variant &p_right, Variant *r_ret, bool &r_valid) {
 		const A &a = *VariantGetInternalPtr<A>::get_ptr(&p_left);
 		const B &b = *VariantGetInternalPtr<B>::get_ptr(&p_right);
-		*r_ret = a << b;
-		r_valid = true;
+		if (b > 0 && b <= 62 && a >= 0) {
+			*r_ret = a << b;
+			r_valid = true;
+		} else {
+			*r_ret = "Invalid parameter when shifting.";
```

Comment:
Maybe use a more informative message, something like "Negative numbers can't be shifted by more than 62 (otherwise, it's unsafe)".

I ask for some help to come up with a better written, and possibly more concise, message.

Reply:
Maybe:
```
"Invalid operands for bit shifting. Only positive operands are supported, and the right hand side operand cannot be higher than 62."
```

That being said, I don't think if it's worth it to limit to `b <= 62`.
It will still fail if you do `2 << 62` with a signed int, or `4 << 62` with an unsigned it. Only `1 << 62` would be safe.

IMO a note in the documentation should be sufficient to make users aware that bitshifting can lead to wrapping around if they exceed the number of bits of the used type. Maybe we can have some GDScript warnings for that eventually too, but I wouldn't prevent the operation if we can't guarantee that we handle all cases.

So I'd suggest leaving it to only signessness checks, and the error:
```
Invalid operands for bit shifting. Only positive operands are supported.
```

Reply:
I think the message @akien-mga proposes is more accurate and user-friendly than the one in the PR.

The one in the PR says the reason for the error is undefined behavior, but that's too an obscure term for most users. In addition, the fact that it's UB in C++ doesn't necessarily mean that it's in GDScript. GDScript could implement a defined behavior, only that we won't and we just want not to support shifts with negatives involved.

The PR is fine to me besides that.